x509util/extensions.go

```diff
@@ -500,3 +533,170 @@ func (s *SerialNumber) UnmarshalJSON(data []byte) error {
+
+		switch san.Type {
+
+		case EmailType:
+			rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeEmail, Class: 2, Bytes: []byte(san.Value)})
+		case DNSType:
+			rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeDNS, Class: 2, Bytes: []byte(san.Value)})
+		case X400AddressType:
+			fallthrough
+		case DirectoryNameType:
+			fallthrough
+		case EDIPartyNameType:
+			return nil, fmt.Errorf("unimplemented SAN type %s", san.Type)
+		case URIType:
+			rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeURI, Class: 2, Bytes: []byte(san.Value)})
+		case IPType:
+			if rawIP := net.ParseIP(san.Value); rawIP != nil {
+				ip := rawIP.To4()
+				if ip == nil {
+					ip = rawIP
+				}
+				rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeIP, Class: 2, Bytes: ip})
+			}
+		case RegisteredIDType:
+			if san.Value != "" {
+				oid, err := parseObjectIdentifier(san.Value)
+				if err == nil {
+					rawBytes, _ := asn1.MarshalWithParams(oid, fmt.Sprintf("tag:%d", nameTypeRegisteredID))
+					rawValues = append(rawValues, asn1.RawValue{FullBytes: rawBytes})
+				}
+			}
+		default:
+			if san.Type != "" {
+				// if san.Type is a valid OID, we assume it is an OtherName
+				oid, err := parseObjectIdentifier(san.Type)
+				if err == nil {
+					var otherNameValue interface{}
+
+					// the OtherName value can be any type depending on the OID
+					// ASN supports a great number of formats (https://www.openssl.org/docs/man1.0.2/man3/ASN1_generate_nconf.html),
+					// but golang's asn1 lib supports much fewer -- for now support anything the golang asn1 marshaller supports
+
+					// The default type is printable, but if the value is prefixed with a type, use that
+					var valueType = "printable"
+					var sanValue = san.Value
+					var rawBytes []byte
+
+					if strings.Contains(san.Value, ";") {
+						valueType = strings.Split(san.Value, ";")[0]
+						sanValue = san.Value[len(valueType)+1:]
+					}
+
+					switch valueType {
+					case "int":
+						var i int
+						i, err = strconv.Atoi(sanValue)
+						if err != nil {
+							return nil, fmt.Errorf("invalid int value for int-typed SAN OtherName %s", san.Type)
+						}
+						rawBytes, err = asn1.Marshal(i)
+					case "oid":
+						var oidVal asn1.ObjectIdentifier
+						oidVal, err = parseObjectIdentifier(sanValue)
+						if err != nil {
+							return nil, fmt.Errorf("invalid OID value for OID-typed SAN OtherName %s", san.Type)
+						}
+
+						rawBytes, err = asn1.Marshal(oidVal)
+					case "raw":
+						// the raw type accepts a base64 encoded byte array which is passed unaltered into the ASN
+						// marshaller. By using this type users can add ASN1 datatypes manually into templates
+						// to support some unsupported types like BMPString or HEX types
+						rawBytes, err = base64.StdEncoding.DecodeString(sanValue)
+					case "utf8":
+						fallthrough
+					case "ia5":
+						fallthrough
+					case "numeric":
+						fallthrough
+					case "printable":
```

Comment:
```suggestion
					case "utf8", "ia5", "numeric", "printable":
```

Reply:
I guess `MarshalWithParams` will fail on ia5 and numeric if the sanValue is not valid. We should add tests for this, and for this method.